lib/charms/acme_client_operator/v0/acme_client.py

```diff
@@ -0,0 +1,238 @@
+    def _plugin_config(self):
+        return None
+```
+Charms that leverage this library also need to specify a `provides` relation in their
+`metadata.yaml` file. For example:
+```yaml
+provides:
+  certificates:
+    interface: tls-certificates
+```
+"""
+import abc
+import logging
+import re
+from abc import abstractmethod
+from typing import Dict, List
+from urllib.parse import urlparse
+
+from charms.tls_certificates_interface.v1.tls_certificates import (  # type: ignore[import]
+    CertificateCreationRequestEvent,
+    TLSCertificatesProvidesV1,
+)
+from cryptography import x509
+from cryptography.x509.oid import NameOID
+from ops.charm import CharmBase
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops.pebble import ExecError
+
+# The unique Charmhub library identifier, never change it
+LIBID = "b3c9913b68dc42b89dfd0e77ac57236d"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+class AcmeClient(CharmBase):
+    """Base charm for charms that use the ACME protocol to get certificates.
+
+    This charm implements the tls_certificates interface as a provider.
+    """
+
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, *args, plugin: str):
+
+        super().__init__(*args)
+        self._server = "https://acme-staging-v02.api.letsencrypt.org/directory"
+        self._csr_path = "/tmp/csr.pem"
+        self._certs_path = "/tmp/.lego/certificates/"
+        self._container_name = list(self.meta.containers.values())[0].name
+        container_name_with_underscores = self._container_name.replace("-", "_")
+        self.tls_certificates = TLSCertificatesProvidesV1(self, "certificates")
+        pebble_ready_event = getattr(self.on, f"{container_name_with_underscores}_pebble_ready")
+        self.framework.observe(pebble_ready_event, self._on_acme_client_pebble_ready)
+        self.framework.observe(
+            self.tls_certificates.on.certificate_creation_request,
+            self._on_certificate_creation_request,
+        )
+        self._plugin = plugin
+        self._email = ""
+
+    def _on_acme_client_pebble_ready(self, event):
+        if not self._email:
+            self.unit.status = BlockedStatus("Email address was not provided.")
+            event.defer()
+            return
+        if not self._server:
+            self.unit.status = BlockedStatus("Server address was not provided.")
+            event.defer()
+            return
+        if not self._email_is_valid(self._email):
+            self.unit.status = BlockedStatus("Invalid email address.")
+            event.defer()
+            return
+        if not self._server_is_valid(self._server):
+            self.unit.status = BlockedStatus("Invalid server address.")
+            event.defer()
+            return
```

Comment:
This does not work, because `self._email` and `self._server` are not initialized when `PebbleReady` is fired. We can just look up the configuration instead.

Reply:
Correct, I changed the way we store and check the values, let me know what you think :)

Reply:
This looks good, I will fetch the changes and test it locally.